pkg/shipper/shipper.go

```diff
@@ -94,108 +91,128 @@ func New(
 	dir string,
 	bucket Bucket,
 	lbls func() labels.Labels,
-	gossipMinTimeFn func(mint int64),
 ) *Shipper {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
 	if lbls == nil {
 		lbls = func() labels.Labels { return nil }
 	}
-	if gossipMinTimeFn == nil {
-		gossipMinTimeFn = func(mint int64) {}
-	}
 	return &Shipper{
-		logger:          logger,
-		dir:             dir,
-		bucket:          bucket,
-		labels:          lbls,
-		gossipMinTimeFn: gossipMinTimeFn,
-		metrics:         newMetrics(r),
+		logger:  logger,
+		dir:     dir,
+		bucket:  bucket,
+		labels:  lbls,
+		metrics: newMetrics(r),
 	}
 }
 
-// Sync performs a single synchronization if the local block data with the remote end.
-func (s *Shipper) Sync(ctx context.Context) {
-	names, err := fileutil.ReadDir(s.dir)
+// Timestamps returns the minimum timestamp for which data is available and the highest timestamp
+// of blocks that were successfully uploaded.
+func (s *Shipper) Timestamps() (minTime, maxSyncTime int64, err error) {
+	meta, err := ReadMetaFile(s.dir)
 	if err != nil {
-		level.Warn(s.logger).Log("msg", "read dir failed", "err", err)
+		return 0, 0, errors.Wrap(err, "read shipper meta file")
+	}
+	// Build a map of blocks we already uploaded.
+	hasUploaded := make(map[ulid.ULID]struct{}, len(meta.Uploaded))
+	for _, id := range meta.Uploaded {
+		hasUploaded[id] = struct{}{}
 	}
 
-	var oldestBlockMinTime int64 = math.MaxInt64
-	for _, fn := range names {
-		id, err := ulid.Parse(fn)
-		if err != nil {
-			continue
-		}
-		dir := filepath.Join(s.dir, fn)
+	minTime = math.MaxInt64
+	maxSyncTime = math.MinInt64
 
-		fi, err := os.Stat(dir)
-		if err != nil {
-			level.Warn(s.logger).Log("msg", "open file failed", "err", err)
-			continue
-		}
-		if !fi.IsDir() {
-			continue
+	s.iterBlockMetas(func(m *block.Meta) error {
+		if m.MinTime < minTime {
+			minTime = m.MinTime
 		}
-		minTime, err := s.sync(ctx, id, dir)
-		if err != nil {
-			level.Error(s.logger).Log("msg", "shipping failed", "dir", dir, "err", err)
-			continue
+		if _, ok := hasUploaded[m.ULID]; ok && m.MaxTime > maxSyncTime {
+			maxSyncTime = m.MaxTime
 		}
+		return nil
+	})
+	return minTime, maxSyncTime, nil
+}
 
-		if minTime < oldestBlockMinTime || oldestBlockMinTime == math.MaxInt64 {
-			oldestBlockMinTime = minTime
+// Sync performs a single synchronization if the local block data with the remote end.
```

Comment:
Comment is bit obsolete: It is not if the local block data is on the remote end, it is "if I uploaded that already"